src/service_control_client_impl_quota_test.cc

```diff
@@ -0,0 +1,372 @@
+
+const char kRequest1[] = R"(
+service_name: "library.googleapis.com"
+allocate_operation {
+  operation_id: "operation-1"
+  method_name: "methodname"
+  consumer_id: "consumerid"
+  quota_metrics {
+    metric_name: "metric_first"
+    metric_values {
+      int64_value: 1
+    }
+  }
+  quota_metrics {
+    metric_name: "metric_second"
+    metric_values {
+      int64_value: 1
+    }
+  }
+  quota_mode: NORMAL
+}
+service_config_id: "2016-09-19r0"
+)";
+
+const char kSuccessResponse1[] = R"(
+operation_id: "operation-1"
+quota_metrics {
+  metric_name: "serviceruntime.googleapis.com/api/consumer/quota_used_count"
+  metric_values {
+    labels {
+      key: "/quota_name"
+      value: "metric_first"
+    }
+    int64_value: 1
+  }
+  metric_values {
+    labels {
+      key: "/quota_name"
+      value: "metric_first"
+    }
+    int64_value: 1
+  }
+}
+service_config_id: "2016-09-19r0"
+)";
+
+const char kErrorResponse1[] = R"(
+operation_id: "a197c6f2-aecc-4a31-9744-b1d5aea4e4b4"
+allocate_errors {
+  code: RESOURCE_EXHAUSTED
+  subject: "user:integration_test_user"
+}
+)";
+
+}  // namespace
+
+class ServiceControlClientImplQuotaTest : public ::testing::Test {
+ public:
+  void SetUp() {
+    ASSERT_TRUE(TextFormat::ParseFromString(kRequest1, &quota_request1_));
+    ASSERT_TRUE(
+        TextFormat::ParseFromString(kSuccessResponse1, &pass_quota_response1_));
+    ASSERT_TRUE(
+        TextFormat::ParseFromString(kErrorResponse1, &error_quota_response1_));
+  }
+
+  AllocateQuotaRequest quota_request1_;
+  AllocateQuotaResponse pass_quota_response1_;
+  AllocateQuotaResponse error_quota_response1_;
+
+  MockCheckTransport mock_check_transport_;
+  MockQuotaTransport mock_quota_transport_;
+  MockReportTransport mock_report_transport_;
+};
+
+TEST_F(ServiceControlClientImplQuotaTest,
+       TestNonCachedQuotaWithStoredCallback) {
+  ServiceControlClientOptions options(CheckAggregationOptions(0, 500, 1000),
+                                      QuotaAggregationOptions(0, 500),
+                                      ReportAggregationOptions(0, 500));
+
+  options.check_transport = mock_check_transport_.GetFunc();
+  options.report_transport = mock_report_transport_.GetFunc();
+
+  int callbackCalledCount = 0;
+  options.quota_transport = [this, &callbackCalledCount](
+      const AllocateQuotaRequest& request, AllocateQuotaResponse* response,
+      TransportDoneFunc on_done) {
+    response = &pass_quota_response1_;
+
+    callbackCalledCount++;
+    on_done(Status::OK);
+  };
+
+  std::unique_ptr<ServiceControlClient> client_ =
+      CreateServiceControlClient(kServiceName, kServiceConfigId, options);
+
+  Status done_status = Status::UNKNOWN;
+  AllocateQuotaResponse quota_response;
+
+  client_->Quota(quota_request1_, &quota_response,
+                 [&done_status](Status status) { done_status = status; });
+  ASSERT_TRUE(TextFormat::ParseFromString(kSuccessResponse1, &quota_response));
```

Comment:
you should verify quota_response is the same as pass_quota_response1 since it is inpace,  both done_stauts, and quota_response are assigned and ready for check